tpm2/test/src/cipher/aes_keywrap_test.c

```diff
@@ -0,0 +1,539 @@
+        else
+        {
+          step++;
+          C_str_len = strlen(buffer) - 4;  // strip leading 'C = ' sub-string
+          memcpy(C_str, buffer + 4, C_str_len * sizeof(char));
+          while ((C_str_len > 0) &&
+                 ((C_str[C_str_len - 1] == '\n') ||
+                  (C_str[C_str_len - 1] == '\r')))
+          {
+            C_str[--C_str_len] = '\0';  // strip any trailing '\n' or '\r'
+          }
+        }
+      }
+  
+      else if (strncmp(buffer, "FAIL", 4) == 0)
+      {
+        // found expected 'FAIL' result, but incomplete vector - reset
+        if (step != 3)
+        {
+          step = 1;
+          P_str_len = 0;
+          C_str_len = 0;
+        }
+
+        // found expected 'FAIL' result (should be in place of 'P' component)
+        else
+        {
+          step++;
+          pass_result = false;
+        }
+      }
+  
+      // found line in vector file that is not in one of the formats parsed by
+      // this function - reset (start over, looking for next vector)
+      else
+      {
+        P_str_len = 0;
+        C_str_len = 0;
+        step = 1;
+      }
+    }
+  }
+
+  // reaching step 4 means a vector has been successfully parsed
+  if (step == 4)
+  {
+    // use parsed results to populate output parameters
+    convert_HexString_to_ByteArray((char **) K_vec,
+                                             K_str,
+                                             K_str_len);
+    *K_vec_len = K_str_len / 2;  // 2 hex chars map to a byte of key
+    convert_HexString_to_ByteArray((char **) P_vec,
+                                             P_str,
+                                             P_str_len);
+    *P_vec_len = P_str_len / 2;  // 2 hex chars map to a byte of key
+    convert_HexString_to_ByteArray((char **) C_vec,
+                                             C_str,
+                                             C_str_len);
+    *C_vec_len = C_str_len / 2;  // 2 hex chars map to a byte of key
+    *expect_pass = pass_result;
+  }
+
+  // clean-up allocated memory
+  free(K_str);
+  free(P_str);
+  free(C_str);
+
+  // if while loop exit due to EOF, return unsuccessful result
+  if (step != 4)
+  {
+    return 1;
+  }
+
+  // normal termination
+  return 0;
+}
+
+//---------------------- AES Key Wrap Cipher Tests ---------------------------
+
+//----------------------------------------------------------------------------
+// test_aes_keywrap_parameters()
+//----------------------------------------------------------------------------
+void test_aes_keywrap_parameters(void)
+{
+  unsigned char* key = NULL;
+  int key_len = 0;
+  
+  unsigned char* inData = NULL;
+  size_t inData_len = 0;
+
+  unsigned char* outData = NULL;
+  size_t outData_len = 0;
+
+  // Test failure on null key
+  inData = malloc(16);
+  inData_len = 16;
+  key_len = 16;
+  CU_ASSERT(aes_keywrap_3394nopad_encrypt(key, key_len, inData, inData_len, &outData, &outData_len) == 1);
```

Comment:
I think after these failure checks you can also verify that `outData == NULL` and that `outDataLen == 0`. The failure conditions in encrypt/decrypt don't always `NULL` `outData`, but the input sanity checks happen before it gets allocated. This applies to all the tests in this function.

Reply:
I think that I added a "CU_ASSERT((outData == NULL) && (outData_len == 0))" statement after all of the tests for out-of-limit input parameter handling.